### ISSUE TITLE
BaseTestPredictorsMaml.Run_Test() is broken

### ISSUE THREAD

**Issue by mstfbl** [dotnet/machinelearning#4789]
I am trying to test a previously saved LightGBM model to see if it outputs expected baseline figures. However, the baseline that `Run_Test()` generates is as follows:

> ***** Unexpected failure. Please refer to https://aka.ms/MLNetIssue to file an issue with details *****
> 
> ***** Error log has been saved to 'C:\Users\mubal\AppData\Local\Temp\TLC\Error_20200206_193056_51208d31-0f5c-4ff9-b7a8-7f853d81fc14.log', please refer to https://aka.ms/MLNetIssue to file an issue with details *****
> 
> ===== Begin detailed dump =====
> 
> (1) Unexpected exception: Assert failed: Assertion Failed
> 
> Expected: True
> 
> Actual:   False, 'Xunit.Sdk.TrueException'
> 
>    at Xunit.Assert.True(Nullable`1 condition, String userMessage) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 95
> 
>    at Microsoft.ML.Internal.Internallearn.Test.GlobalBase.AssertHandler(String msg, IExceptionContext ectx) in C:\Users\mubal\Documents\GitHub\machinelearning\test\Microsoft.ML.TestFramework\GlobalBase.cs:line 71
> 
>    at Microsoft.ML.Runtime.Contracts.DbgFailCore(String msg, IExceptionContext ctx) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Core\Utilities\Contracts.cs:line 781
> 
>    at Microsoft.ML.Runtime.Contracts.DbgFail(IExceptionContext ctx) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Core\Utilities\Contracts.cs:line 790
> 
>    at Microsoft.ML.Runtime.Contracts.Assert(Boolean f) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Core\Utilities\Contracts.cs:line 839
> 
>    at Microsoft.ML.Runtime.ComponentCatalog.ParseArguments(IHostEnvironment env, Object args, String settings, String name) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Core\ComponentModel\ComponentCatalog.cs:line 1067
> 
>    at Microsoft.ML.Runtime.ComponentCatalog.TryCreateInstance[TRes](IHostEnvironment env, Type signatureType, TRes& result, String name, String options, Object[] extra) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Core\ComponentModel\ComponentCatalog.cs:line 1036
> 
>    at Microsoft.ML.Runtime.ComponentCatalog.TryCreateInstance[TRes,TSig](IHostEnvironment env, TRes& result, String name, String options, Object[] extra) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Core\ComponentModel\ComponentCatalog.cs:line 993
> 
>    at Microsoft.ML.Tools.Maml.MainCore(IHostEnvironment env, String args, Boolean alwaysPrintStacktrace) in C:\Users\mubal\Documents\GitHub\machinelearning\src\Microsoft.ML.Maml\MAML.cs:line 129
> 
> ====== End detailed dump =====

My code where this happens is:

https://github.com/dotnet/machinelearning/blob/0ea84a6f3fb2cde29f70ec638c3f06317028225e/test/Microsoft.ML.Predictor.Tests/TestPredictors.cs#L489-L511

I've confirmed that my model and and baseline files exist, and that my model is valid.

Perhaps this issue was known to previous ML.NET developers as Run_Test() and its dependent Run_TrainSaveTest() are not used anywhere in the codebase. However, this is the only way (as far as I know) i n which one can test the baseline output of a model with `TestDatasets` and `TestLearners`.

Because Run_Test() is broken, Run_TrainSaveTest() is also broken.

PS: I discovered this while trying to add a unit test in my PR #4695  to check the run-time behavior of LightGBM does not change by modifying the flags used by LightGBM with CursOpt.AllFeatures, and that this change does not affect the features extracted during validation.

**Comment by antoniovs1029** [comment]
**TL;DR** It seems to me that `BaseTestPredictorsMaml.RunTest()` isn't broken, and that the problem in here is in the way you're trying to use the `breastCancerPipe` object. To quickly fix your problem try commenting out (or removing) the  `mamlExtraSettings = new[] { "cache-" },` line in

https://github.com/dotnet/machinelearning/blob/52c0dc66d3af8953c8692fd546609fd4514861f9/test/Microsoft.ML.TestFrameworkCommon/Datasets.cs#L97

By doing this I was able to get the correct baseline output file `LightGBM-Test-breast-cancer-out.txt`, and then by copying that file into the appropriate folder (namely, `machinelearning\test\BaselineOutput\Common\LightGBMBinary`) I was able to run successfully your `LightGBMPreviousModelBaselineTest`.

Note: When you update your PR to add the new test, I think it would be best not to change the existing `breastCancerPipe` object, but rather create a new object like `breastCancerPipe2` (or whatever descriptive name you might decide) which is exactly the same as `breastCancerPipe`, but without the `mamlExtraSettings`.

-----------------------------
### Why this seem to work?
If I run your test as-is and then look into the baseline output file `LightGBM-Test-breast-cancer-out.txt` I do get the same content that you've posted in your first post. If I instead use the debugger to step through your test, the same file gets a different content:

```
Unrecognized command line argument 'cache'
Usage For 'Test':
featureColumn=<string>              Column to use for features Default value:'Features' (short form feat)
[...]
```
I don't know why the difference in content happens when you Run a Test instead of Debugging it. But at least this other output file is more informative, as it says that there's an argument called 'cache' which isn't recognized.

This error message seems to happen because the MAML command that is constructed in `Microsoft.ML.Run.Tests.BaseTestPredictors.Run()` takes its arguments from the `breastCancerPipe` object, and then the command is passed as "args" to the `Microsoft.ML.RunTests.TestCommandBase.MainForTest()` method. At this point, if I run your test with the breastCancerPipe object unmodified, then "args" contains the following command:

`"Test data=C:\\Users\\anvelazq\\source\\repos\\antoniovs1029\\machinelearning\\test\\data\\breast-cancer.txt seed=1 loader=Text{sparse- col=Attr:TX:6 col=Label:0 col=Features:1-5,6,7-9} cache- dout={C:\\Users\\anvelazq\\source\\repos\\antoniovs1029\\machinelearning\\bin\\AnyCPU.Debug\\Microsoft.ML.Predictor.Tests\\netcoreapp2.1\\TestOutput\\LightGBMBinary\\LightGBM-Test-breast-cancer.txt} in=C:\\Users\\anvelazq\\Desktop\\issue4695\\model.zip"`

Notice the "cache-" argument in there.

As the output file that I got while debugging suggests, the "cache-" argument isn't available for the `Test` command, and so by removing it from the `breastCancerPipe` object, this stops causing problems in your test.

Also notice that the `BaseTestPredictorsMaml.Run_Test()` method uses "Cmd.Test" in herem which determines that the "Test" command is going to be used by MAML:
https://github.com/dotnet/machinelearning/blob/52c0dc66d3af8953c8692fd546609fd4514861f9/test/Microsoft.ML.TestFramework/BaseTestPredictorsMaml.cs#L421-L422

In contrast, `Run_TrainTest()` (which is used, for example, by `LightGBMClassificationTest()`), uses "Cmd.TrainTest", which then determines the "Train Test" command for MAML.

My assumption is that the "Train Test" command does accept the "cache-" argument, whereas the Test command doesn't, and that's why you couldn't use the same `breatCancerPipe` object in your `Run_Test()` method call.

**Comment by mstfbl** [comment]
Thank you so much @antoniovs1029 ! It's definitely very interesting how this behavior changes when its run in debug vs. release mode.
